[Cocoa] Add tests to validate the behavior of SpatialAudioExperience

jernoble opened the pull request:
#### 50eeda809aa991027d04d2d93738622c8bf90464
<pre>
[Cocoa] Add tests to validate the behavior of SpatialAudioExperience
<a href="https://bugs.webkit.org/show_bug.cgi?id=292195">https://bugs.webkit.org/show_bug.cgi?id=292195</a>
&lt;<a href="https://rdar.apple.com/problem/150212519">rdar://problem/150212519</a>&gt;

Reviewed by Andy Estes.

After adopting -setIntendedSpatialAudioExperience:, make the behavior testable
by adding a facility to pass messages up from the GPU process to the Internals
object. This facility can be re-used for other messages and other objects:

- MessageClientForTesting: An abstract interface which allows introspected objects
  to pass generic messages to clients for testing purposes.
- MessageForTesting: A simple message object containing a type and data string.
- EventTargetForTesting: A concrete EventTarget instantiated by Internals
- CustomEvent: A concrete Event subclass which can be initialized with a payload.

These classes are used by HTMLMediaElement, MediaPlayer, MediaPlayerInterface,
MediaPlayerPrivateRemote and RemoteMediaPlayer to provide a channel between
concrete MediaPlayerInterface classes like MediaPlayerPrivateAVFoundationObjC
and MediaPlayerPrivateMediaSourceAVFObjC, and the Internals object in the
WebContent process.

And this in turn enables tests like SpatialAudioExperience.NoWindow,
SpatialAudioExperience.WindowWithWindowScene and SpatialAudioExperience.AudioOnly.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::addMessageClientForTesting):
(WebCore::HTMLMediaElement::removeMessageClientForTesting):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/MessageClientForTesting.h: Added.
(WebCore::AggregateMessageClientForTesting::create):
(WebCore::AggregateMessageClientForTesting::addClient):
(WebCore::AggregateMessageClientForTesting::removeClient):
(WebCore::AggregateMessageClientForTesting::isEmpty const):
* Source/WebCore/platform/MessageForTesting.h: Added.
* Source/WebCore/platform/MessageTargetForTesting.h: Added.
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::loadWithNextMediaEngine):
(WebCore::MediaPlayer::setMessageClientForTesting):
(WebCore::MediaPlayer::protectedMessageClientForTesting const):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::setMessageClientForTesting):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateSpatialTrackingLabel):
* Source/WebCore/testing/EventForTesting.h: Added.
(WebCore::EventForTesting::create):
(WebCore::EventForTesting::data const):
(WebCore::EventForTesting::EventForTesting):
* Source/WebCore/testing/EventForTesting.idl: Added.
* Source/WebCore/testing/EventTargetForTesting.cpp: Added.
(WebCore::EventTargetForTesting::create):
(WebCore::EventTargetForTesting::~EventTargetForTesting):
(WebCore::EventTargetForTesting::EventTargetForTesting):
(WebCore::EventTargetForTesting::sendInternalMessage):
* Source/WebCore/testing/EventTargetForTesting.h: Added.
* Source/WebCore/testing/EventTargetForTesting.idl: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::addInternalEventTarget):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setHasMessageClientForTesting):
(WebKit::RemoteMediaPlayerProxy::sendInternalMessage):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(WKSLinearMediaPlayer.isImmersiveVideo):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setMessageClientForTesting):
(WebKit::MediaPlayerPrivateRemote::sendInternalMessage):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SpatialAudioExperience.mm: Added.
(-[SpatialAudioExperienceMessageHandler userContentController:didReceiveScriptMessage:]):
(TestWebKitAPI::autoplayingInternalsConfiguration):
(TestWebKitAPI::TEST(SpatialAudioExperience, NoWindow)):
(TestWebKitAPI::TEST(SpatialAudioExperience, WindowWithWindowScene)):
(TestWebKitAPI::TEST(SpatialAudioExperience, AudioOnly)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/spatial-audio-experience-with-audio.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/spatial-audio-experience-with-video.html: Added.

Canonical link: <a href="https://commits.webkit.org/295336@main">https://commits.webkit.org/295336@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3171cc114fb61bf09887c7684b8822445d95a99e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24446 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109948 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55408 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106774 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24837 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32992 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79535 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107740 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19312 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94529 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59842 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19080 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12606 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54786 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88777 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12653 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112345 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31899 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23462 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88610 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90755 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88234 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22496 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33119 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10899 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27212 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31824 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37177 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31616 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34957 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33175 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/html/HTMLMediaElement.h:
```diff
@@ -95,6 +96,8 @@ class Event;
 class HTMLTrackElement;
 class InbandTextTrackPrivate;
+class MessageClientForTesting;
+class AggregateMessageClientForTesting;
```
Nit: mis-sorted

aestes commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -2071,6 +2074,17 @@ const Logger& MediaPlayer::mediaPlayerLogger()
 
+void MediaPlayer::setMessageClientForTesting(WeakPtr<MessageClientForTesting> internalMessageClient)
+{
+    m_internalMessageClient = internalMessageClient;
```
```suggestion
    m_internalMessageClient = WTFMove(internalMessageClient);
```

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -4092,6 +4094,10 @@ void determineChangedTracksFromNewTracksAndOldItems(MediaSelectionGroupAVFObjC*
         ALWAYS_LOG(LOGIDENTIFIER, "Setting spatialAudioExperience: ", spatialAudioExperienceDescription(experience.get()));
         [m_avPlayer setIntendedSpatialAudioExperience:experience.get()];
+
+        if (RefPtr internalMessageClient = player->protectedMessageClientForTesting())
```
Nit: I've learned that this is an incorrect use of the `protected` pattern. Apparently in cases where we store to a local variable we should call a `messageClientForTesting()` function that returns a `MessageClientForTesting*`.

(Which begs the question: when would you ever have a `protectedFoo()` that returns `RefPtr`, since you always want to null-check those)

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -1819,6 +1821,10 @@ void getSupportedTypes(HashSet<String>& types) const final
         ALWAYS_LOG(LOGIDENTIFIER, "Setting spatialAudioExperience: ", spatialAudioExperienceDescription(experience.get()));
         [m_synchronizer setIntendedSpatialAudioExperience:experience.get()];
+
+        if (RefPtr internalMessageClient = player->protectedMessageClientForTesting())
```
Nit: see above.

aestes commented on Source/WebCore/testing/EventTargetForTesting.cpp:
```diff
@@ -0,0 +1,61 @@
+
+EventTargetForTesting::~EventTargetForTesting()
+{
+    if (RefPtr<MessageTargetForTesting> messageTarget { m_messageTarget.get() })
```
Nit: does this work?
```suggestion
    if (RefPtr messageTarget { m_messageTarget.get() })
```

jernoble replied:
There was a reason for this at the time, but I may have fixed the underlying problem preventing this from working correctly.

aestes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
```diff
@@ -139,6 +139,7 @@ enum LinearMediaPlayerErrors: Error {
     var isImmersiveVideo: Bool {
         get { swiftOnlyData.isImmersiveVideo }
         set {
+            Logger.linearMediaPlayer.log("\(#function) \(newValue)")
```
Is this supposed to be part of this PR?

aestes commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -1887,6 +1888,25 @@ void MediaPlayerPrivateRemote::sceneIdentifierDidChange()
 
+void MediaPlayerPrivateRemote::setMessageClientForTesting(WeakPtr<MessageClientForTesting> client)
+{
+    m_internalMessageClient = client;
```
```suggestion
    m_internalMessageClient =WTFMove(client);
```

aestes commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
```diff
@@ -64,6 +64,16 @@
+            argument = "--no-timeout"
+            isEnabled = "YES">
+         </CommandLineArgument>
+      </CommandLineArguments>
```
Can we revert the changes to this file?

aestes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SpatialAudioExperience.mm:
```diff
@@ -0,0 +1,142 @@
+- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message
+{
+    if (!_expectedBody)
+        return;
```
Nit: this is unnecessary

jernoble committed:
[Cocoa] Add tests to validate the behavior of SpatialAudioExperience
https://bugs.webkit.org/show_bug.cgi?id=292195
<rdar://problem/150212519>

Reviewed by Andy Estes.

After adopting -setIntendedSpatialAudioExperience:, make the behavior testable
by adding a facility to pass messages up from the GPU process to the Internals
object. This facility can be re-used for other messages and other objects:

- MessageClientForTesting: An abstract interface which allows introspected objects
  to pass generic messages to clients for testing purposes.
- MessageForTesting: A simple message object containing a type and data string.
- EventTargetForTesting: A concrete EventTarget instantiated by Internals
- CustomEvent: A concrete Event subclass which can be initialized with a payload.

These classes are used by HTMLMediaElement, MediaPlayer, MediaPlayerInterface,
MediaPlayerPrivateRemote and RemoteMediaPlayer to provide a channel between
concrete MediaPlayerInterface classes like MediaPlayerPrivateAVFoundationObjC
and MediaPlayerPrivateMediaSourceAVFObjC, and the Internals object in the
WebContent process.

And this in turn enables tests like SpatialAudioExperience.NoWindow,
SpatialAudioExperience.WindowWithWindowScene and SpatialAudioExperience.AudioOnly.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::addMessageClientForTesting):
(WebCore::HTMLMediaElement::removeMessageClientForTesting):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/MessageClientForTesting.h: Added.
(WebCore::AggregateMessageClientForTesting::create):
(WebCore::AggregateMessageClientForTesting::addClient):
(WebCore::AggregateMessageClientForTesting::removeClient):
(WebCore::AggregateMessageClientForTesting::isEmpty const):
* Source/WebCore/platform/MessageForTesting.h: Added.
* Source/WebCore/platform/MessageTargetForTesting.h: Added.
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::loadWithNextMediaEngine):
(WebCore::MediaPlayer::setMessageClientForTesting):
(WebCore::MediaPlayer::protectedMessageClientForTesting const):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::setMessageClientForTesting):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateSpatialTrackingLabel):
* Source/WebCore/testing/EventForTesting.h: Added.
(WebCore::EventForTesting::create):
(WebCore::EventForTesting::data const):
(WebCore::EventForTesting::EventForTesting):
* Source/WebCore/testing/EventForTesting.idl: Added.
* Source/WebCore/testing/EventTargetForTesting.cpp: Added.
(WebCore::EventTargetForTesting::create):
(WebCore::EventTargetForTesting::~EventTargetForTesting):
(WebCore::EventTargetForTesting::EventTargetForTesting):
(WebCore::EventTargetForTesting::sendInternalMessage):
* Source/WebCore/testing/EventTargetForTesting.h: Added.
* Source/WebCore/testing/EventTargetForTesting.idl: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::addInternalEventTarget):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setHasMessageClientForTesting):
(WebKit::RemoteMediaPlayerProxy::sendInternalMessage):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(WKSLinearMediaPlayer.isImmersiveVideo):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setMessageClientForTesting):
(WebKit::MediaPlayerPrivateRemote::sendInternalMessage):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SpatialAudioExperience.mm: Added.
(-[SpatialAudioExperienceMessageHandler userContentController:didReceiveScriptMessage:]):
(TestWebKitAPI::autoplayingInternalsConfiguration):
(TestWebKitAPI::TEST(SpatialAudioExperience, NoWindow)):
(TestWebKitAPI::TEST(SpatialAudioExperience, WindowWithWindowScene)):
(TestWebKitAPI::TEST(SpatialAudioExperience, AudioOnly)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/spatial-audio-experience-with-audio.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/spatial-audio-experience-with-video.html: Added.

Canonical link: https://commits.webkit.org/295336@main
